### PR TITLE
perf(p3-goldilocks): add SVE2 vectorization

### PR DESCRIPTION
**Pull request by Al-Kindi-0**
Three additions to the aarch64 packed path, gated on cfg(target_feature = "sve2") except the NEON hardening:
- vector-domain modular mul (z-register mul/umulh, vsraq-based reduction), used by the packed Mul impl on SVE2 targets
- delayed-reduction dot product: 128-bit products accumulate unreduced in per-lane lo/hi/lo-wrap counters, one exact 192-bit reduction at the end; exact for any N (carry bound debug-asserted), covered by edge-value and wrap-saturation tests
- Poseidon2 external/internal round asm with compiler-visible scheduling; all predication pinned to two 64-bit lanes (svptrue_pat_b64(SV_VL2)), so the code is safe at any SVE vector length; full clobber declarations
- NEON round helpers canonicalize the constant operand before add/sub, closing a second-order overflow on adversarial residues (regression test included); scalar helpers document their input-domain contracts

### CODE REVIEW

File: goldilocks/src/aarch64_neon/poseidon2_asm.rs [hunk 1]
```diff
@@ -1389,6 +1395,85 @@ pub fn internal_permute_neon_w12(state: &mut [uint64x2_t; 12], constants: &[u64]
...
 }
 
+// Scalar Goldilocks multiply after 0xPolygonZero/plonky2's
+// `poseidon_goldilocks_neon.rs` (MIT/Apache-2.0). Unlike `mul_asm`, whose
+// single asm block is opaque to the scheduler, these helpers are mostly plain
+// Rust, so LLVM interleaves the four s0 chains of the internal rounds. The
+// wraparound correction after the `lo - (hi >> 32)` step is a branch rather
+// than two ALU ops: it is taken with probability ~2^-32.
+// Structurally this mirrors `goldilocks.rs`'s `reduce128` with aarch64
+// micro-tunings (asm barrier, `umull` w-form); kept local pending upstream discussion.
```
Comment (Nashtare) [inline goldilocks/src/aarch64_neon/poseidon2_asm.rs]:
```suggestion
// micro-tunings (asm barrier, `umull` w-form).
```

File: goldilocks/src/aarch64_neon/poseidon2_asm.rs [hunk 2]
```diff
@@ -1389,6 +1395,85 @@ pub fn internal_permute_neon_w12(state: &mut [uint64x2_t; 12], constants: &[u64]
     state[0] = s0;
 }
 
+// Scalar Goldilocks multiply after 0xPolygonZero/plonky2's
+// `poseidon_goldilocks_neon.rs` (MIT/Apache-2.0). Unlike `mul_asm`, whose
```
Comment (Nashtare) [inline goldilocks/src/aarch64_neon/poseidon2_asm.rs]:
```suggestion
// `poseidon_goldilocks_neon.rs`. Unlike `mul_asm`, whose
```

attribution already done workspace wise

File: goldilocks/src/aarch64_neon/poseidon2_asm.rs [hunk 3]
```diff
@@ -1524,6 +1609,179 @@ pub fn internal_permute_neon<const WIDTH: usize>(
...
+    }
+}
+
+/// x^7 S-box layer over the whole state, SVE2. Six-element blocks where the
+/// width allows (six independent multiply chains cover the vector-mul latency
+/// better than four), four-element blocks otherwise.
+#[cfg(target_feature = "sve2")]
+#[inline(always)]
+unsafe fn sbox_neon<const WIDTH: usize>(state: &mut [uint64x2_t; WIDTH]) {
+    const { assert!(WIDTH % 6 == 0 || WIDTH % 4 == 0) }
```
Comment (Nashtare) [inline goldilocks/src/aarch64_neon/poseidon2_asm.rs]: not a blocker, but this const assert feels a bit weird as standalone given that the non-sve2 counterpart doesn't enforce it. Though not an issue in practice

File: goldilocks/src/aarch64_neon/poseidon2_asm.rs [hunk 4]
```diff
@@ -3191,3 +3449,106 @@ mod tests {
...
+    use rand::rngs::SmallRng;
+    use rand::{RngExt, SeedableRng};
+
+    use super::*;
+    use crate::Goldilocks;
+
+    /// Raw operand patterns stressing every wraparound correction: field
+    /// extremes, the epsilon window, and non-canonical values up to u64::MAX.
+    /// The ops under test accept any u64 residue.
+    const EDGE: [u64; 8] = [0, 1, (1 << 32) - 1, 1 << 32, 1 << 63, P - 1, P, u64::MAX];
```
Comment (Nashtare) [inline goldilocks/src/aarch64_neon/poseidon2_asm.rs]: nit: defined twice

File: goldilocks/src/aarch64_neon/poseidon2_asm.rs [hunk 5]
```diff
@@ -1524,6 +1609,179 @@ pub fn internal_permute_neon<const WIDTH: usize>(
...
+/// width allows (six independent multiply chains cover the vector-mul latency
+/// better than four), four-element blocks otherwise.
+#[cfg(target_feature = "sve2")]
+#[inline(always)]
+unsafe fn sbox_neon<const WIDTH: usize>(state: &mut [uint64x2_t; WIDTH]) {
+    const { assert!(WIDTH % 6 == 0 || WIDTH % 4 == 0) }
+    unsafe {
+        if WIDTH % 6 == 0 {
+            for chunk in state.chunks_exact_mut(6) {
+                sbox6_sve2(chunk.try_into().unwrap());
```
Comment (Nashtare) [inline goldilocks/src/aarch64_neon/poseidon2_asm.rs]: wouldn't `as_chunks_mut()` work here? To remove the unnecessary unwrap

File: goldilocks/src/aarch64_neon/poseidon2_asm.rs [hunk 6]
```diff
@@ -1524,6 +1609,179 @@ pub fn internal_permute_neon<const WIDTH: usize>(
...
+            }
+        } else {
+            for chunk in state.chunks_exact_mut(4) {
+                sbox4_sve2(chunk.try_into().unwrap());
```
Comment (Nashtare) [inline goldilocks/src/aarch64_neon/poseidon2_asm.rs]: same here

File: goldilocks/src/aarch64_neon/utils.rs [hunk 7]
```diff
@@ -5,7 +5,7 @@ use core::arch::asm;
 use super::packing::PackedGoldilocksNeon;
 use crate::{Goldilocks, P};
 
-const EPSILON: u64 = P.wrapping_neg(); // 2^32 - 1
+pub(super) const EPSILON: u64 = P.wrapping_neg(); // 2^32 - 1
```
Comment (Nashtare) [inline goldilocks/src/aarch64_neon/utils.rs:8]: Can also be reused in `packing.rs` module

File: goldilocks/src/aarch64_neon/poseidon2_asm.rs (hunk 3 above)
Comment (Nashtare) [inline goldilocks/src/aarch64_neon/poseidon2_asm.rs]: also to silence clippy:


```suggestion
    const { assert!(WIDTH.is_multiple_of(6) || WIDTH .is_multiple_of(4)) }
```

**Review by Nashtare — COMMENTED** [review]
Thanks! I'm getting ~9% improvement on Poseidon2 with width 12!
Although as we're adding a new target this would need a dedicated step in CI to make sure it compiles fine at least

File: goldilocks/src/aarch64_neon/poseidon2_asm.rs [hunk 8]
```diff
@@ -1524,6 +1609,179 @@ pub fn internal_permute_neon<const WIDTH: usize>(
...
+
+/// x^7 S-box layer over the whole state, SVE2. Six-element blocks where the
+/// width allows (six independent multiply chains cover the vector-mul latency
+/// better than four), four-element blocks otherwise.
+#[cfg(target_feature = "sve2")]
+#[inline(always)]
+unsafe fn sbox_neon<const WIDTH: usize>(state: &mut [uint64x2_t; WIDTH]) {
+    const { assert!(WIDTH % 6 == 0 || WIDTH % 4 == 0) }
+    unsafe {
+        if WIDTH % 6 == 0 {
```
Comment (Nashtare) [inline goldilocks/src/aarch64_neon/poseidon2_asm.rs]: same here for `is_multiple_of` and below

File: goldilocks/src/aarch64_neon/poseidon2_asm.rs (hunk 3 above)
Comment (Al-Kindi-0) [reply goldilocks/src/aarch64_neon/poseidon2_asm.rs]: Kept the assert, added the missing rationale

**Review by Nashtare — APPROVED** [review]
LGTM, thanks again!
